### PR TITLE
cleanup go e2e tests and perform improvements

### PR DESCRIPTION
**Pull request by camilamacedo86**
**Description of the change:**
- Update the go e2e test to have the suffix `_test` which has [special meaning for the Go compiler](https://github.com/golang/go/issues/25223).
- Cleanup the test in order to make all following the same standard
- Ensure that the test are done using the makefile targets scaffolds as we suggest in our docs. 
- Add tests to ensure the metrics, run the project locally and the bundle sdk features
- fix the target in the doc https://master.sdk.operatorframework.io/docs/olm-integration/generation/#package-manifests-format. Users should use `$(KUSTOMIZE)` instead of kustomize directly.
- Fix Prometheus typo issue 

**Motivation for the change:**

- maintainability, readability and reusability
- ensure that our e2e tests are using the targets commands as we suggested to the users in the docs

NOTE: The first goal is to keep all tests following the same standard. However, it still has a hall for improvements that ought to make horizontally in follow-ups.

### CODE REVIEW

**Comment by camilamacedo86** [comment]
Hi @estroz, 

> I’m not so sure we want to do this, at least not yet. Go e2e tests, while they could test more functionality, are fine as-is for v1.0. We can prioritize improving them after release. 

I do not agree with first because I do not see why it affects the v1.0 release. Also, we are applying here the changes/discussions made in the helm and I do not think that the test is fine now because they are not using the makefile as we suggest to in the docs and we are unable to run them locally with any env for example.

**Comment by camilamacedo86** [comment]
@joelanford wdyt? Can we move forward with this one?

**Comment by joelanford** [comment]
> fix the target in the doc master.sdk.operatorframework.io/docs/olm-integration/generation/#package-manifests-format. Users should use $(KUSTOMIZE) instead of kustomize directly.

This is the only thing from this PR that I think would be better to get in before 1.0. Everything else is nice to have, but _can_ merge before 1.0 (if we get time to review, etc.)

I agree with @estroz that 1.0 _blockers_ are the priority at the moment.

**Comment by camilamacedo86** [comment]
Removing the /hold since has I understand according to above comments that have no reason to hold that besides we are not prioritizing it for 1.0.0.

/hold cancel

File: hack/tests/e2e-go.sh
Comment (estroz) [inline hack/tests/e2e-go.sh:10]: Why are any of these changes necessary?

Comment (camilamacedo86) [reply hack/tests/e2e-go.sh:10]: when we run the script locally if we do not `make install` first it will not get the changes
My idea/suggestion proposed in another pr was to keep the make install in the script instead of Travis.

File: test/e2e/e2e_go_olm_test.go
Comment (camilamacedo86) [inline test/e2e/e2e_go_olm_test.go:58]:
```suggestion
```

Comment (camilamacedo86) [reply test/e2e/e2e_go_olm_test.go:58]: I need to remove it because it is not required

File: test/e2e/e2e_go_suite_test.go
Comment (jberkhahn) [inline test/e2e/e2e_go_suite_test.go:148]: Nice use of testing suite.

File: test/e2e/e2e_go_cluster_test.go
Comment (jberkhahn) [inline test/e2e/e2e_go_cluster_test.go:70]: So, shouldn't most of these conditionals be using ginkgo-style checks? Like this should be Expect(len(podNames)).To(Equal(1)) instead of manually bubbling up an error?

**Comment by jberkhahn** [comment]
I think this looks nice to split this out like this. A couple spots could be cleaned up a bit to be more ginkgo-esque in style, though.

File: test/e2e/e2e_go_cluster_test.go
Comment (camilamacedo86) [reply test/e2e/e2e_go_cluster_test.go:70]: Since the other tests are with the same check. I think it would be nice be done horizontally to all after we get Ansible merged as well. WDYT?

Comment (camilamacedo86) [reply test/e2e/e2e_go_cluster_test.go:70]: See; https://github.com/operator-framework/operator-sdk/issues/3771

Comment (jberkhahn) [reply test/e2e/e2e_go_cluster_test.go:70]: :+1:

File: test/e2e/e2e_go_suite_test.go
Comment (jmccormick2001) [inline test/e2e/e2e_go_suite_test.go:44]: just a type-o catch, all instances of 'Promethues' should be changed to 'Prometheus'

Comment (camilamacedo86) [reply test/e2e/e2e_go_suite_test.go]: fixed.

File: test/e2e/e2e_go_cluster_test.go
Comment (joelanford) [reply test/e2e/e2e_go_cluster_test.go]: Cross-post from https://github.com/operator-framework/operator-sdk/issues/3771#issuecomment-691482813:

> I think this can be improved further to `Expect(podNames).To(HaveLen(1))`

**Review by jmrodri — APPROVED** [review]
/lgtm

File: test/internal/packagemanifests.go
Comment (asmacdo) [inline test/internal/packagemanifests.go:64]:
```suggestion
// More info:  https://v1-0-x.sdk.operatorframework.io/docs/olm-integration/generation/#package-manifests-formats
```

File: test/internal/packagemanifests.go
Comment (asmacdo) [inline test/internal/packagemanifests.go:81]:
```suggestion
// More info:  https://v1-0-x.sdk.operatorframework.io/docs/olm-integration/generation/#package-manifests-formats
```

**Review by asmacdo — APPROVED** [review]

File: test/internal/packagemanifests.go
Comment (joelanford) [inline test/internal/packagemanifests.go:46]: Why is this not included in the NonGoFragment?

File: test/internal/packagemanifests.go
Comment (joelanford) [inline test/internal/packagemanifests.go:37]: Why do we specify `--interactive=false` here, but not for Go?

**Review by joelanford — CHANGES_REQUESTED** [review]

File: test/internal/packagemanifests.go
Comment (camilamacedo86) [reply test/internal/packagemanifests.go:37]: We have a target for GO and for NonGO projects. It does NOT equal for both. See: https://v1-0-x.sdk.operatorframework.io/docs/olm-integration/generation/#package-manifests-formats

Comment (camilamacedo86) [reply test/internal/packagemanifests.go]: Because, the packagemanifests will call the bundle if it was not generated yet. 

```
$ make packagemanifests 
operator-sdk generate kustomize manifests -q

Display name for the operator (required): 
> 
```

File: test/internal/packagemanifests.go
Comment (camilamacedo86) [reply test/internal/packagemanifests.go:29]: Good catcher

File: test/e2e-go/e2e_go_olm_test.go
Comment (joelanford) [inline test/e2e-go/e2e_go_olm_test.go:105]: We talked about not running these tests as part of the Go e2e. Can we remove from here and do a follow-up to add them to the `test/integration` tests?

File: test/e2e-go/e2e_go_suite_test.go
Comment (joelanford) [inline test/e2e-go/e2e_go_suite_test.go:107]: Ditto from above. Also, I don't think we want to keep that comment around since the custom tests are just examples. We'll never scaffold a real scorecard config file with these examples present.

File: test/e2e-go/e2e_go_suite_test.go
Comment (joelanford) [inline test/e2e-go/e2e_go_suite_test.go:131]: IMO, this would add unnecessary complexity to the Go e2e for very limited value. We already test `InstrumentedEnqueueRequestForObject` in unit tests in `operator-lib` and in e2e tests for Helm and Ansible operators.

These e2e tests are meant to ensure that the scaffolding, packaging, and execution of Go projects works. Testing `InstrumentedEnqueueRequestForObject` here would be similar to testing controller-runtime's `EnqueueRequestsFromMapFunc` handler here. Both handlers are unrelated to this test.

File: test/e2e-go/e2e_go_cluster_test.go
Comment (joelanford) [inline test/e2e-go/e2e_go_cluster_test.go:190]: Ditto. I don't think the added complexity of adding SDK metrics with a custom handler has much value for the scope of these e2e tests.

File: test/internal/packagemanifests.go
Comment (joelanford) [inline test/internal/packagemanifests.go:69]: This is brittle. If we change the packagemanifests target to something slightly different, this replacement will silently fail.

Instead, what if we make the fragment a format string that we pass to `fmt.Sprintf()`. e.g.

```go
makefilePackagemanifestsFragment := `
# Options for "packagemanifests".
ifneq ($(origin FROM_VERSION), undefined)
PKG_FROM_VERSION := --from-version=$(FROM_VERSION)
endif
ifneq ($(origin CHANNEL), undefined)
PKG_CHANNELS := --channel=$(CHANNEL)
endif
ifeq ($(IS_CHANNEL_DEFAULT), 1)
PKG_IS_DEFAULT_CHANNEL := --default-channel
endif
PKG_MAN_OPTS ?= $(FROM_VERSION) $(PKG_CHANNELS) $(PKG_IS_DEFAULT_CHANNEL)
# Generate package manifests.
packagemanifests: %s
	operator-sdk generate kustomize manifests -q --interactive=false
	cd config/manager && $(KUSTOMIZE) edit set image controller=$(IMG)
	$(KUSTOMIZE) build config/manifests | operator-sdk generate packagemanifests -q --version $(VERSION) $(PKG_MAN_OPTS)
`

pmDeps := []string{"kustomize"}
if strings.HasPrefix(c.Layout, "go") {
	pmDeps = append(pmDeps, "manifests")
}
fragment := fmt.Sprintf(makefilePackagemanifestsFragment, strings.Join(pmDeps, " "))
```

**Review by joelanford — CHANGES_REQUESTED** [review]
One more round now that I had more time to look at it.

Looks great overall, but a I have a few more comments and suggestions.

File: test/e2e-go/e2e_go_cluster_test.go
Comment (camilamacedo86) [reply test/e2e-go/e2e_go_cluster_test.go:53]: It is exactly it is other tests. I would like to keep all equals. If we decide to change then we can do it horizontally.

File: test/e2e-go/e2e_go_cluster_test.go
Comment (camilamacedo86) [reply test/e2e-go/e2e_go_cluster_test.go:56]: ditto above

File: test/e2e-go/e2e_go_olm_test.go
Comment (camilamacedo86) [reply test/e2e-go/e2e_go_olm_test.go:105]: I think we need to discuss that before to do a follow. These tests were added via tasks planned in the sprint which means that was discussed before to get done. Also, I think we should keep them. See that when we split the repos will make sense we have only the e2e tests to ensure the SDK features ( scorecard and olm ) on this repo. 

Also, see that the `test/integration` is not following the standards define so far and with them, we will NOT test it end to end and then, we will not able to ensure that the commands work with the project layout and with the makefile targets suggested in the docs.

File: test/e2e-go/e2e_go_cluster_test.go
Comment (camilamacedo86) [reply test/e2e-go/e2e_go_cluster_test.go:53]: see: https://github.com/operator-framework/operator-sdk/issues/3878

File: test/e2e-go/e2e_go_cluster_test.go
Comment (camilamacedo86) [reply test/e2e-go/e2e_go_cluster_test.go:60]: see: #3878

File: test/e2e-go/e2e_go_olm_test.go
Comment (camilamacedo86) [reply test/e2e-go/e2e_go_olm_test.go:105]: See; https://github.com/operator-framework/enhancements/pull/44/files#diff-0af5b52afc5f9db2b08e97937d991b40R233

File: test/e2e-go/e2e_go_suite_test.go
Comment (camilamacedo86) [reply test/e2e-go/e2e_go_suite_test.go:107]: It is fine we can remove the init. However, I understood that the idea was in a long term scaffold the examples by default. Then, it would remove the mock here as well. It is was the only reason for the comment.

File: test/e2e-go/e2e_go_suite_test.go
Comment (camilamacedo86) [reply test/e2e-go/e2e_go_suite_test.go:131]: Agree.

File: test/internal/packagemanifests.go
Comment (camilamacedo86) [reply test/internal/packagemanifests.go]: It is good always check. You are right here.

File: test/e2e-go/e2e_go_olm_test.go
Comment (joelanford) [reply test/e2e-go/e2e_go_olm_test.go:105]: Ah I see now that this is just copied over from the existing file. Seems reasonable to keep for now, but we should add a TODO comment to split the scorecard custom test testing out.

We should also discuss if there's really a need for the custom test example at all. After all, the basic and olm tests are probably reasonable examples on their own.

File: test/e2e-go/e2e_go_suite_test.go
Comment (joelanford) [reply test/e2e-go/e2e_go_suite_test.go]: I think that idea was to add some boilerplate `kuttl` tests as part of the `create api` command.

File: test/internal/packagemanifests.go
Comment (joelanford) [inline test/internal/packagemanifests.go]: Nit: string comparison and replacement seems more complex than the `fmt.Sprintf` idea, and it is still prone to accidental breakage. At least now the test would fail, but it seems more efficient and less error prone to build the string instead of search/replace it. 

With this, it is possible that we'll make a change and not see that we accidentally broke something until 20-30 minutes into a CI run.

**Review by joelanford — APPROVED** [review]
Looks good overall. One more nit about the packagemanifests templating.

Otherwise
/lgtm

File: test/e2e-go/e2e_go_olm_test.go
Comment (camilamacedo86) [reply test/e2e-go/e2e_go_olm_test.go:105]: I will push soon the PR to split the scorecard.

**Pull request merged**
